src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -102,6 +109,31 @@ internal TElement[] ToArray(int minIdx, int maxIdx)
             return array;
         }
 
+        internal List<TElement> ToList(int minIdx, int maxIdx)
+        {
+            Buffer<TElement> buffer = new Buffer<TElement>(source);
+            int count = buffer.count;
+            if (count <= minIdx) return new List<TElement>();
+            if (count <= maxIdx) maxIdx = count - 1;
+            if (minIdx == maxIdx) return new List<TElement>(1) { GetEnumerableSorter().ElementAt(buffer.items, count, minIdx) };
+            int[] map = SortedMap(buffer, minIdx, maxIdx);
+            List<TElement> list = new List<TElement>(maxIdx - minIdx + 1);
+            while (minIdx <= maxIdx)
+            {
+                list.Add(buffer.items[map[minIdx]]);
+                ++minIdx;
+            }
+            return list;
+        }
+
+        internal int GetCount(int minIdx, int maxIdx, bool onlyIfCheap)
+        {
+            int count = GetCount(onlyIfCheap);
+            if (count <= 0) return count;
+            if (count <= minIdx) return 0;
```

Comment:
Right now, none.

I had this as a simple `Count`, but you raised a concern at https://github.com/dotnet/corefx/pull/5777#discussion_r51340315 about it being assumed to be cheap and used as the basis of another optimisation.

I'd already an idea of using it as such, but testing if it really was cheap (I was originally going to have a property for that, but as you pointed out that would mean an extra interface call), so while I'm not doing that yet (this PR is large as it is) I am including this mechanism for avoiding a count calculation that isn't cheap here.

(My plan for actually using this is a `Select` that acts on partitions, and will in turn allow for the list-based Select iterators to in turn implement IPartition and so have almost all of the possible advantages of implementing `IList<T>` in them happen without such an implementation, at least as far as time complexity goes).